server/client.go

```diff
@@ -2212,32 +2226,30 @@ func (c *client) parseSub(argo []byte, noForward bool) error {
 	arg := make([]byte, len(argo))
 	copy(arg, argo)
 	args := splitArg(arg)
-	var (
-		subject []byte
-		queue   []byte
-		sid     []byte
-	)
+	sub := &subscription{client: c}
 	switch len(args) {
 	case 2:
-		subject = args[0]
-		queue = nil
-		sid = args[1]
+		sub.subject = args[0]
+		sub.queue = nil
+		sub.sid = args[1]
 	case 3:
-		subject = args[0]
-		queue = args[1]
-		sid = args[2]
+		sub.subject = args[0]
+		sub.queue = args[1]
+		sub.sid = args[2]
 	default:
 		return fmt.Errorf("processSub Parse Error: '%s'", arg)
 	}
 	// If there was an error, it has been sent to the client. We don't return an
 	// error here to not close the connection as a parsing error.
-	c.processSub(subject, queue, sid, nil, noForward)
+	c.processSub(sub, noForward)
 	return nil
 }
 
-func (c *client) processSub(subject, queue, bsid []byte, cb msgHandler, noForward bool) (*subscription, error) {
-	// Create the subscription
-	sub := &subscription{client: c, subject: subject, queue: queue, sid: bsid, icb: cb}
+func (c *client) createSub(subject, queue, sid []byte, cb msgHandler) *subscription {
+	return &subscription{client: c, subject: subject, queue: queue, sid: sid, icb: cb}
+}
+
+func (c *client) processSub(sub *subscription, noForward bool) (*subscription, error) {
```

Comment:
This is because a subscription on "foo" may already exist, so if I call processSub() with a new subscription object but for the same subject/sid, then the existing one is returned (and some of the sub's field are updated - under the client lock. For mqtt it means the qos and other).

Reply:
ok then why do you need to build the sub beforehand?

Reply:
This has 2 advantages: reduce the number of parameters that we would have to had to processSub(). Granted we could create a suboption struct that we pass, but this is really equivalent to passing a sub object, even better if happen to not exist so we don't create the suboption object AND the subscription.
The second is that I need the subscription object to "add" to it retained message prior to be added to the sublist and possibly starting to receive incoming messages.

Reply:
What additional params do you need to pass?
Can you not ass in retained message post this without causing issues?

Reply:
There is qos, but not only that, the subscription has a `mqtt` field that holds the JS consumer it is attached to. I want this to be set before I call processSub() because I want to make sure that this is set prior to add the sub to sublist. You can't set things to the sub after processSub() returns, it is simply not safe. Remember that I had to fix races where you would set sub.icb after calling processSub().
For the retained message, I don't think in the spec they say that they must be sent before any incoming messages, although my experience with some mqtt brokers is that they are sent first.

Reply:
ok that makes sense. Maybe we need to change the name now though from `procesSub`